src/shape_shifter_access.cpp

```diff
@@ -0,0 +1,55 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+// topic_tools::ShapeShifter does allow access to data buffer
+// Unfortunately this means, we can't access data without performing a potentially very
+// expensize message by serializing it to a stream
+// Since ROS1 is no longer in developement adding a new accessor function is not possible.
+// However since ShapeShifter will not change, we can include it by redefining the meaning
+// of private
+
+// Include everything that shape_shifter.h does
+// This will define header include guards so when shape shifter is included,
+// their definitions don't change and cause odd problems.
+#include "ros/ros.h"
+#include "ros/console.h"
+#include "ros/assert.h"
+#include <vector>
+#include <string>
+#include <string.h>
+#include <ros/message_traits.h>
+#include "topic_tools/macros.h"
+
+// Redefine the meaning of private for just new stuff in shape_shifter header
```

Comment:
What's going on here?

Reply:
ShapeShifter does not allow access to its internal buffer (unlike the ROS2 SerializedMessage).   For the ROS1 use cases there is no need for this because functions are provided to serialize data from/to a ROS1 stream.   However, I want to convert the data stored inside a SerializedMessage into a concrete ROS2 message type without making another copy.

Since ROS1 is no longer maintained, it's not  possible to request a change to ROS that provides an accessor function to ShapeShifter's internal buffer.

Typically classes will make "internal" data protected/private when there is no obvious reason to allow access.   This provides flexibility to change in the internal representation in the future.  Since ROS1 is not getting any update, this no longer matters.

If this seems too ugly, I can always define another type like ShapeShifter that can act like a container for any generic ROS1 type.

Reply:
One thought I just had is I could just have generic function take a const reference to a buffer (std::vector<uint8_t>) and let external called figure out how they want get access to the buffer.
Like I said before it is possible for someone to create something very similar to the ShapeShifter to subscribe to "generic" ROS1 streams.